server/cluster/cluster.go

```diff
@@ -1051,6 +1058,65 @@ func (c *RaftCluster) RemoveStore(storeID uint64, physicallyDestroyed bool) erro
 	return err
 }
 
+func (c *RaftCluster) checkReplicaBeforeOfflineStore(storeID uint64) error {
+	upStores := c.getUpStores()
+	expectUpStoresNum := len(upStores) - 1
+	// When placement rules feature is enabled. It is hard to determine required replica count precisely.
+	if !c.opt.IsPlacementRulesEnabled() && expectUpStoresNum < c.opt.GetMaxReplicas() {
+		return fmt.Errorf("can not remove store %d since there are no extra up store has enough space to accommodate the extra replica（%d）", storeID, c.opt.GetMaxReplicas())
```

Comment:
We have another place to give the warning which is duplicated with this, I think we can remove it since it's better to prevent the operation.

Reply:
As we discussed, It is ok we check the `placementRules` twice ^^